02-express-tutorial/07-params-query.js

```diff
@@ -0,0 +1,102 @@
+        const {id, name, image} = product;
+        return {id, name, image};
+    });
+
+    res.json(newProducts);
+});
+
+// //We can use "route" parameters here instead of hard coding this param
+// app.get("/api/products/1", (req, res) => {
+//     const singleProduct = products.find((product) => product.id === 1);
+
+//     res.json(singleProduct)
+// });
+
+// Using "route" parameters:
+app.get("/api/products/:productID", (req, res) => {
+    // console.log(req);
+    // console.log(req.params);
+    // "productID" is set up as number in the JSON but we have to convert it
+    // to string to use it in the route parameter here. Hence, Number(productID).
+    // With this, users can use request and route parameters to access data.
+    const { productID } = req.params;
+    const singleProduct = products.find((product) => product.id === Number(productID)
+    );
+
+    //If the query params doesn't exist in the resource on the url queried, then do this:
+    if(!singleProduct) {
+        return res.status(404).send("Product Does Not Exist")
+    }
+
+    console.log(singleProduct);
+    res.json(singleProduct);
+});
+
+app.get("/api/products/:productID/reviews/:reviewID", (req, res) => {
+    console.log(req.params);
+    res.send("hello world");
+});
+
+//  The "query string" parameters are also called the "url" parameters.
+//  This is a way for the client to send small amount of information 
+//  back to query the server using the "url".
+
+//  Whatever comes after the "?" is not technically part of the url. 
+app.get("/api/v1/query", (req, res) => {
+    //  In order to access those query parameters, use "req.query".
+    //  So in the browser, type in "http://localhost:5000/api/v1/query?name=daniel&id=1". 
+    //  Or using the "search" and the "limit" parameters, type in "http://localhost:5000/api/v1/query?search=a&limit=2",
+    //  to seacrh for names with "a", and a max (limit) of 2 items should be returned.
+    //  The server will respond with the hard coded string: "hello world". 
+    // console.log(req.query);
+    
+
+    // To setup the query functionality:
+    // First create a new instance(copy) of those products(sortedProducts) using the spread operator.
+    const { search, limit } = req.query;
+    let sortedProducts = [...products];
+
+    if (search) {
+        sortedProducts = sortedProducts.filter((product) => {
+            //  We will get errors in the server console if we don't explicitly include "return" 
+            //  in our "if" statements, except it is the last statement in the "if" block. 
+            //  This is because JS will try to send two responses for the same request. 
+            //  It doesnt termnate the previous statement if "return" isn't used. 
+            //  Hence, JS try to send both the current response and the next response at the same time.
+            //  Meanwhile, in JS, you can only have one response per request. 
+            return product.name.startsWith(search); 
```

Comment:
To clarify what's happening here, you are returning a boolean (true/false) if the product in the list starts with the search term. That's separate from what eventually gets returned as the response to the request. (This return statement is actually _within_ the `.filter` method, which isolates it from returning out of the `if` statement.)

Reply:
Thanks for this explanation.

Appreciate!